File: tools/source/sdks/cs-sdk/src/GameAgentEngineClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace GameAgentEngine.SDK;

public sealed class GameAgentEngineClient
{
    private readonly HttpClient _http;
    private readonly JsonSerializerOptions _json;
    private readonly string _baseUrl;

    public GameAgentEngineClient(string baseUrl, string apiKey, HttpClient? httpClient = null, JsonSerializerOptions? jsonSerializerOptions = null)
    {
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            throw new ArgumentException("baseUrl is required", nameof(baseUrl));
        }

        if (string.IsNullOrWhiteSpace(apiKey))
        {
            throw new ArgumentException("apiKey is required", nameof(apiKey));
        }

        _baseUrl = baseUrl.TrimEnd('/');
        _http = httpClient ?? new HttpClient();
        _json = jsonSerializerOptions ?? JsonDefaults.Create();

        _http.DefaultRequestHeaders.Remove("X-API-Key");
        _http.DefaultRequestHeaders.Add("X-API-Key", apiKey);
        if (!_http.DefaultRequestHeaders.Accept.Any(h => h.MediaType == "application/json"))
        {
            _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }
    }

    public Task<JsonElement?> HealthAsync(CancellationToken cancellationToken = default) =>
        SendAsync<JsonElement?>(HttpMethod.Get, "/health", null, null, cancellationToken);

    public Task<VersionInfo?> GetVersionAsync(CancellationToken cancellationToken = default) =>
        SendAsync<VersionInfo>(HttpMethod.Get, "/api/v1/version", null, null, cancellationToken);

    public Task<InvokeResponse?> InvokeAsync(InvokeRequest request, CancellationToken cancellationToken = default) =>
        SendAsync<InvokeResponse>(HttpMethod.Post, "/api/v1/invoke", request, null, cancellationToken);

    public Task<InvokeResponse?> InterpretPlayerInputAsync(PlayerInputInterpretRequest request, CancellationToken cancellationToken = default) =>
        SendAsync<InvokeResponse>(HttpMethod.Post, "/api/v1/player/input/interpret", request, null, cancellationToken);

    public Task<TickResponse?> AdvanceTickAsync(string worldId, TickAdvanceRequest request, CancellationToken cancellationToken = default) =>
        SendAsync<TickResponse>(HttpMethod.Post, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/ticks/advance", request, null, cancellationToken);

    public Task<WorldSettings?> GetWorldSettingsAsync(string worldId, CancellationToken cancellationToken = default) =>
        SendAsync<WorldSettings>(HttpMethod.Get, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/settings", null, null, cancellationToken);

    public Task<WorldSettings?> SetWorldSettingsAsync(string worldId, object settings, CancellationToken cancellationToken = default) =>
        SendAsync<WorldSettings>(HttpMethod.Put, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/settings", settings, null, cancellationToken);

    public Task<StateComponentsResponse?> GetStateComponentsAsync(string worldId, CancellationToken cancellationToken = default) =>
        SendAsync<StateComponentsResponse>(HttpMethod.Get, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/state-components", null, null, cancellationToken);

    public Task<StateComponentResponse?> GetStateComponentAsync(string worldId, string componentType, CancellationToken cancellationToken = default) =>
        SendAsync<StateComponentResponse>(HttpMethod.Get, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/state-components/{Uri.EscapeDataString(componentType)}", null, null, cancellationToken);

    public Task<StateComponentResponse?> PutStateComponentAsync(string worldId, string componentType, object payload, CancellationToken cancellationToken = default) =>
        SendAsync<StateComponentResponse>(HttpMethod.Put, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/state-components/{Uri.EscapeDataString(componentType)}", payload, null, cancellationToken);

    public Task<TimelinesResponse?> GetTimelinesAsync(string worldId, int? limit = null, CancellationToken cancellationToken = default) =>
        SendAsync<TimelinesResponse>(HttpMethod.Get, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/timelines{BuildQuery(("limit", limit))}", null, null, cancellationToken);

    public Task<LatestTimelineResponse?> GetLatestTimelineAsync(string worldId, CancellationToken cancellationToken = default) =>
        SendAsync<LatestTimelineResponse>(HttpMethod.Get, $"/api/v1/worlds/{Uri.EscapeDataString(worldId)}/timelines/latest", null, null, cancellationToken);

    public Task<List<InferenceLog>?> GetLogsAsync(InferenceLogQuery? query = null, CancellationToken cancellationToken = default) =>
        SendAsync<List<InferenceLog>>(HttpMethod.Get, $"/api/v1/logs{BuildQuery(query)}", null, null, cancellationToken);

    public Task<DebugTraceList?> GetDebugTracesAsync(string worldId, int limit = 20, CancellationToken cancellationToken = default) =>
        SendAsync<DebugTraceList>(HttpMethod.Get, $"/debug/traces{BuildQuery(("world_id", worldId), ("limit", limit))}", null, null, cancellationToken);

    public Task<RuntimeTaskListResponse?> ListRuntimeTasksAsync(string? category = null, string? status = null, int limit = 20, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskListResponse>(HttpMethod.Get, $"/api/v1/runtime/tasks{BuildQuery(("category", category), ("status", status), ("limit", limit))}", null, null, cancellationToken);

    public Task<RuntimeTaskListResponse?> ListPendingRuntimeTasksAsync(string consumer, int limit = 20, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskListResponse>(HttpMethod.Get, $"/api/v1/runtime/tasks/pending{BuildQuery(("consumer", consumer), ("limit", limit))}", null, null, cancellationToken);

    public Task<RuntimeTaskEnvelope?> GetRuntimeTaskAsync(string taskId, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskEnvelope>(HttpMethod.Get, $"/api/v1/runtime/tasks/{Uri.EscapeDataString(taskId)}", null, null, cancellationToken);

    public Task<RuntimeTaskEnvelope?> ClaimRuntimeTaskAsync(string taskId, string consumer, string leaseOwner, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskEnvelope>(HttpMethod.Post, "/api/v1/runtime/tasks/claim", new { task_id = taskId, consumer, lease_owner = leaseOwner }, null, cancellationToken);

    public Task<RuntimeTaskEnvelope?> StartRuntimeTaskAsync(string taskId, string leaseToken, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskEnvelope>(HttpMethod.Post, "/api/v1/runtime/tasks/start", new { task_id = taskId, lease_token = leaseToken }, null, cancellationToken);

    public Task<JsonElement?> HeartbeatRuntimeTaskAsync(string taskId, string leaseToken, CancellationToken cancellationToken = default) =>
        SendAsync<JsonElement?>(HttpMethod.Post, "/api/v1/runtime/tasks/heartbeat", new { task_id = taskId, lease_token = leaseToken }, null, cancellationToken);

    public Task<JsonElement?> ReleaseRuntimeTaskAsync(string taskId, string leaseToken, string? errorMessage = null, CancellationToken cancellationToken = default) =>
        SendAsync<JsonElement?>(HttpMethod.Post, "/api/v1/runtime/tasks/release", new { task_id = taskId, lease_token = leaseToken, error_message = errorMessage ?? string.Empty }, null, cancellationToken);

    public Task<RuntimeTaskEnvelope?> RequeueRuntimeTaskAsync(string taskId, int retryDelayMs = 0, string? errorMessage = null, CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskEnvelope>(HttpMethod.Post, "/api/v1/runtime/tasks/requeue", new { task_id = taskId, retry_delay_ms = retryDelayMs, error_message = errorMessage ?? string.Empty }, null, cancellationToken);

    public Task<RuntimeTaskStatsEnvelope?> GetRuntimeTaskStatsAsync(CancellationToken cancellationToken = default) =>
        SendAsync<RuntimeTaskStatsEnvelope>(HttpMethod.Get, "/api/v1/runtime/tasks/stats", null, null, cancellationToken);

    public Task<CallbackResponse?> ActionCallbackAsync(string callbackId, string status, object result, string? requestId = null, CancellationToken cancellationToken = default)
    {
        Dictionary<string, string>? headers = null;
        if (!string.IsNullOrWhiteSpace(requestId))
        {
            headers = new Dictionary<string, string>
            {
                ["X-Callback-Request-Id"] = requestId,
            };
        }

        return SendAsync<CallbackResponse>(HttpMethod.Post, "/api/v1/actions/callback", new { callback_id = callbackId, status, result }, headers, cancellationToken);
    }

    private async Task<T?> SendAsync<T>(HttpMethod method, string path, object? body, IDictionary<string, string>? extraHeaders, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, _baseUrl + path);
        if (extraHeaders is not null)
        {
            foreach (var header in extraHeaders)
            {
                request.Headers.Remove(header.Key);
                request.Headers.Add(header.Key, header.Value);
            }
        }

        if (body is not null)
        {
            var json = JsonSerializer.Serialize(body, _json);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        using var response = await _http.SendAsync(request, cancellationToken).ConfigureAwait(false);
        var text = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new GameAgentEngineException($"HTTP {(int)response.StatusCode} {method} {path}", (int)response.StatusCode, path, text);
        }

        if (string.IsNullOrWhiteSpace(text))
        {
            return default;
        }

        return JsonSerializer.Deserialize<T>(text, _json);
    }

    private static string BuildQuery(params (string Key, object? Value)[] pairs)
    {
        var items = new List<string>();
        foreach (var (key, value) in pairs)
        {
            if (value is null)
            {
                continue;
            }

            var text = value switch
            {
                string s when string.IsNullOrWhiteSpace(s) => null,
                string s => s,
                _ => Convert.ToString(value, System.Globalization.CultureInfo.InvariantCulture),
            };

            if (string.IsNullOrWhiteSpace(text))
            {
                continue;
            }

            items.Add($"{Uri.EscapeDataString(key)}={Uri.EscapeDataString(text)}");
        }

        return items.Count == 0 ? string.Empty : "?" + string.Join("&", items);
    }

    private static string BuildQuery(InferenceLogQuery? query)
    {
        if (query is null)
        {
            return string.Empty;
        }

        return BuildQuery(
            ("world_id", query.WorldId),
            ("node_id", query.NodeId),
            ("task_type", query.TaskType),
            ("category", query.Category),
            ("event_name", query.EventName),
            ("execution_mode", query.ExecutionMode),
            ("request_id", query.RequestId),
            ("round", query.Round),
            ("limit", query.Limit),
            ("offset", query.Offset));
    }
}

internal static class JsonDefaults
{
    public static JsonSerializerOptions Create() => new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = false,
    };
}

public sealed class GameAgentEngineException : Exception
{
    public GameAgentEngineException(string message, int statusCode, string path, string? responseText) : base(message)
    {
        StatusCode = statusCode;
        Path = path;
        ResponseText = responseText;
    }

    public int StatusCode { get; }
    public string Path { get; }
    public string? ResponseText { get; }
}

File: tools/source/sdks/cpp-sdk/src/game_agent_engine_client.cpp
#include "game_agent_engine_client.hpp"

#include <sstream>

GameAgentEngineClient::GameAgentEngineClient(std::string baseUrl, std::string apiKey)
    : base_url_(std::move(baseUrl)), api_key_(std::move(apiKey)) {}

std::string GameAgentEngineClient::healthPath() const { return base_url_ + "/health"; }
std::string GameAgentEngineClient::versionPath() const { return base_url_ + "/api/v1/version"; }
std::string GameAgentEngineClient::invokePath() const { return base_url_ + "/api/v1/invoke"; }
std::string GameAgentEngineClient::interpretPlayerInputPath() const { return base_url_ + "/api/v1/player/input/interpret"; }
std::string GameAgentEngineClient::advanceTickPath(const std::string& worldId) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/ticks/advance";
}

std::string GameAgentEngineClient::worldSettingsPath(const std::string& worldId) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/settings";
}

std::string GameAgentEngineClient::stateComponentsPath(const std::string& worldId) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/state-components";
}

std::string GameAgentEngineClient::stateComponentPath(const std::string& worldId, const std::string& componentType) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/state-components/" + componentType;
}

std::string GameAgentEngineClient::timelinesPath(const std::string& worldId, int limit) const {
    std::ostringstream path;
    path << base_url_ << "/api/v1/worlds/" << worldId << "/timelines";
    if (limit > 0) {
        path << "?limit=" << limit;
    }
    return path.str();
}

std::string GameAgentEngineClient::latestTimelinePath(const std::string& worldId) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/timelines/latest";
}

std::string GameAgentEngineClient::logsPath(const std::string& worldId, int limit, int offset, const std::string& taskType) const {
    std::ostringstream path;
    path << base_url_ << "/api/v1/logs?world_id=" << worldId;
    if (limit > 0) path << "&limit=" << limit;
    if (offset > 0) path << "&offset=" << offset;
    if (!taskType.empty()) path << "&task_type=" << taskType;
    return path.str();
}

std::string GameAgentEngineClient::debugTracesPath(const std::string& worldId, int limit) const {
    std::ostringstream path;
    path << base_url_ << "/debug/traces?world_id=" << worldId;
    if (limit > 0) path << "&limit=" << limit;
    return path.str();
}

std::string GameAgentEngineClient::worldPolicyPath(const std::string& worldId) const {
    return base_url_ + "/api/v1/worlds/" + worldId + "/policy";
}
std::string GameAgentEngineClient::pendingTasksPath(const std::string& consumer, int limit) const {
    return base_url_ + "/api/v1/runtime/tasks/pending?consumer=" + consumer + "&limit=" + std::to_string(limit);
}

std::string GameAgentEngineClient::runtimeTasksPath(const std::string& category, const std::string& status, int limit) const {
    std::ostringstream path;
    path << base_url_ << "/api/v1/runtime/tasks?limit=" << limit;
    if (!category.empty()) path << "&category=" << category;
    if (!status.empty()) path << "&status=" << status;
    return path.str();
}

std::string GameAgentEngineClient::runtimeTaskPath(const std::string& taskId) const {
    return base_url_ + "/api/v1/runtime/tasks/" + taskId;
}

std::string GameAgentEngineClient::claimRuntimeTaskPayload(const std::string& taskId, const std::string& consumer, const std::string& owner) const {
    return std::string("{\"task_id\":\"") + taskId + "\",\"consumer\":\"" + consumer + "\",\"lease_owner\":\"" + owner + "\"}";
}

std::string GameAgentEngineClient::startRuntimeTaskPayload(const std::string& taskId, const std::string& leaseToken) const {
    return std::string("{\"task_id\":\"") + taskId + "\",\"lease_token\":\"" + leaseToken + "\"}";
}

std::string GameAgentEngineClient::heartbeatRuntimeTaskPayload(const std::string& taskId, const std::string& leaseToken) const {
    return std::string("{\"task_id\":\"") + taskId + "\",\"lease_token\":\"" + leaseToken + "\"}";
}

std::string GameAgentEngineClient::releaseRuntimeTaskPayload(const std::string& taskId, const std::string& leaseToken, const std::string& errorMessage) const {
    return std::string("{\"task_id\":\"") + taskId + "\",\"lease_token\":\"" + leaseToken + "\",\"error_message\":\"" + errorMessage + "\"}";
}

std::string GameAgentEngineClient::requeueRuntimeTaskPayload(const std::string& taskId, int retryDelayMs, const std::string& errorMessage) const {
    return std::string("{\"task_id\":\"") + taskId + "\",\"retry_delay_ms\":" + std::to_string(retryDelayMs) + ",\"error_message\":\"" + errorMessage + "\"}";
}

std::string GameAgentEngineClient::callbackPayload(const std::string& callbackId, const std::string& status, const std::string& resultJson) const {
    return std::string("{\"callback_id\":\"") + callbackId + "\",\"status\":\"" + status + "\",\"result\":" + resultJson + "}";
}

GameAgentEngineRequest GameAgentEngineClient::healthRequest() const { return {"GET", "/health", ""}; }
GameAgentEngineRequest GameAgentEngineClient::invokeRequest(const std::string& bodyJson) const { return {"POST", "/api/v1/invoke", bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::interpretPlayerInputRequest(const std::string& bodyJson) const { return {"POST", "/api/v1/player/input/interpret", bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::advanceTickRequest(const std::string& worldId, const std::string& bodyJson) const { return {"POST", "/api/v1/worlds/" + worldId + "/ticks/advance", bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::worldSettingsRequest(const std::string& worldId) const { return {"GET", "/api/v1/worlds/" + worldId + "/settings", ""}; }
GameAgentEngineRequest GameAgentEngineClient::setWorldSettingsRequest(const std::string& worldId, const std::string& bodyJson) const { return {"PUT", "/api/v1/worlds/" + worldId + "/settings", bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::stateComponentsRequest(const std::string& worldId) const { return {"GET", "/api/v1/worlds/" + worldId + "/state-components", ""}; }
GameAgentEngineRequest GameAgentEngineClient::stateComponentRequest(const std::string& worldId, const std::string& componentType) const { return {"GET", "/api/v1/worlds/" + worldId + "/state-components/" + componentType, ""}; }
GameAgentEngineRequest GameAgentEngineClient::putStateComponentRequest(const std::string& worldId, const std::string& componentType, const std::string& bodyJson) const { return {"PUT", "/api/v1/worlds/" + worldId + "/state-components/" + componentType, bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::timelinesRequest(const std::string& worldId, int limit) const { return {"GET", timelinesPath(worldId, limit).substr(base_url_.size()), ""}; }
GameAgentEngineRequest GameAgentEngineClient::latestTimelineRequest(const std::string& worldId) const { return {"GET", "/api/v1/worlds/" + worldId + "/timelines/latest", ""}; }
GameAgentEngineRequest GameAgentEngineClient::logsRequest(const std::string& worldId, int limit, int offset, const std::string& taskType) const { return {"GET", logsPath(worldId, limit, offset, taskType).substr(base_url_.size()), ""}; }
GameAgentEngineRequest GameAgentEngineClient::debugTracesRequest(const std::string& worldId, int limit) const { return {"GET", debugTracesPath(worldId, limit).substr(base_url_.size()), ""}; }
GameAgentEngineRequest GameAgentEngineClient::worldPolicyRequest(const std::string& worldId) const { return {"GET", "/api/v1/worlds/" + worldId + "/policy", ""}; }
GameAgentEngineRequest GameAgentEngineClient::setWorldPolicyRequest(const std::string& worldId, const std::string& bodyJson) const { return {"PUT", "/api/v1/worlds/" + worldId + "/policy", bodyJson}; }
GameAgentEngineRequest GameAgentEngineClient::pendingTasksRequest(const std::string& consumer, int limit) const { return {"GET", "/api/v1/runtime/tasks/pending?consumer=" + consumer + "&limit=" + std::to_string(limit), ""}; }
GameAgentEngineRequest GameAgentEngineClient::runtimeTasksRequest(const std::string& category, const std::string& status, int limit) const { return {"GET", runtimeTasksPath(category, status, limit).substr(base_url_.size()), ""}; }
GameAgentEngineRequest GameAgentEngineClient::runtimeTaskRequest(const std::string& taskId) const { return {"GET", "/api/v1/runtime/tasks/" + taskId, ""}; }
GameAgentEngineRequest GameAgentEngineClient::claimRuntimeTaskRequest(const std::string& taskId, const std::string& consumer, const std::string& owner) const { return {"POST", "/api/v1/runtime/tasks/claim", claimRuntimeTaskPayload(taskId, consumer, owner)}; }
GameAgentEngineRequest GameAgentEngineClient::startRuntimeTaskRequest(const std::string& taskId, const std::string& leaseToken) const { return {"POST", "/api/v1/runtime/tasks/start", startRuntimeTaskPayload(taskId, leaseToken)}; }
GameAgentEngineRequest GameAgentEngineClient::heartbeatRuntimeTaskRequest(const std::string& taskId, const std::string& leaseToken) const { return {"POST", "/api/v1/runtime/tasks/heartbeat", heartbeatRuntimeTaskPayload(taskId, leaseToken)}; }
GameAgentEngineRequest GameAgentEngineClient::releaseRuntimeTaskRequest(const std::string& taskId, const std::string& leaseToken, const std::string& errorMessage) const { return {"POST", "/api/v1/runtime/tasks/release", releaseRuntimeTaskPayload(taskId, leaseToken, errorMessage)}; }
GameAgentEngineRequest GameAgentEngineClient::requeueRuntimeTaskRequest(const std::string& taskId, int retryDelayMs, const std::string& errorMessage) const { return {"POST", "/api/v1/runtime/tasks/requeue", requeueRuntimeTaskPayload(taskId, retryDelayMs, errorMessage)}; }
GameAgentEngineRequest GameAgentEngineClient::runtimeTaskStatsRequest() const { return {"GET", "/api/v1/runtime/tasks/stats", ""}; }
GameAgentEngineRequest GameAgentEngineClient::actionCallbackRequest(const std::string& callbackId, const std::string& status, const std::string& resultJson) const { return {"POST", "/api/v1/actions/callback", callbackPayload(callbackId, status, resultJson)}; }
const std::string& GameAgentEngineClient::baseUrl() const { return base_url_; }
const std::string& GameAgentEngineClient::apiKey() const { return api_key_; }
